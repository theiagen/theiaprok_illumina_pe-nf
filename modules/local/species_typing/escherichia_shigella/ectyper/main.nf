process ECTYPER {
    tag "$meta.id"
    label "process_medium"

    container "us-docker.pkg.dev/general-theiagen/biocontainers/ectyper:1.0.0--pyhdfd78af_1"

    input:
    tuple val(meta), path(assembly)
    val o_min_percent_identity
    val h_min_percent_identity
    val o_min_percent_coverage
    val h_min_percent_coverage
    val verify
    val print_alleles

    output:
    tuple val(meta), path("*.tsv"), emit: ectyper_results
    tuple val(meta), path("PREDICTED_SEROTYPE.txt"), emit: ectyper_predicted_serotype_file
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def opid_arg = o_min_percent_identity ? o_min_percent_identity : 90
    def hpid_arg = h_min_percent_identity ? h_min_percent_identity : 95
    def opcov_arg = o_min_percent_coverage ? o_min_percent_coverage : 90
    def hpcov_arg = h_min_percent_coverage ? h_min_percent_coverage : 50
    def verify_arg = verify ? "--verify" : ""
    def alleles_arg = print_alleles ? "-s" : ""
    
    """
    ectyper \\
        -opid ${opid_arg} \\
        -hpid ${hpid_arg} \\
        -opcov ${opcov_arg} \\
        -hpcov ${hpcov_arg} \\
        ${verify_arg} \\
        ${alleles_arg} \\
        --cores ${task.cpus} \\
        --output ./ \\
        --input ${assembly} \\
        ${args}

    # Rename output file
    if [ -f output.tsv ]; then
        mv output.tsv ${prefix}.tsv
    else
        touch ${prefix}.tsv
    fi
    
    # parse ECTyper TSV - extract predicted serotype from column 5
    if [ -f ${prefix}.tsv ] && [ \$(wc -l < ${prefix}.tsv) -gt 1 ]; then
        predicted_serotype=\$(cut -f 5 ${prefix}.tsv | tail -n 1)
    else
        # If the file is empty or does not exist, set a default message
        predicted_serotype="No serotype predicted"
    fi
    echo "\$predicted_serotype" > PREDICTED_SEROTYPE.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ectyper: \$(ectyper --version 2>&1 | sed 's/.*ectyper //; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.tsv
    echo "No serotype predicted" > PREDICTED_SEROTYPE.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ectyper: \$(ectyper --version 2>&1 | sed 's/.*ectyper //; s/ .*\$//')
    END_VERSIONS
    """
}