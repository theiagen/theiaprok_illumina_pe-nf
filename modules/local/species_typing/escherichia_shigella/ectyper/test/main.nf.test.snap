{
    "ectyper - ecoli assembly": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.tsv:md5,45d5b1cc63e5250207cb70126fcdd021"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "PREDICTED_SEROTYPE.txt:md5,d156c404bc17982a9095dc81a706328e"
                ]
            ],
            [
                "versions.yml:md5,c7fe273c583a801fec7101806fc75d7d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T18:51:03.871674454"
    },
    "ectyper - prokaryote assembly - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PREDICTED_SEROTYPE.txt:md5,d4b61d77cb0b810024447774a7d4d141"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c7fe273c583a801fec7101806fc75d7d"
                ],
                "ectyper_predicted_serotype_file": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PREDICTED_SEROTYPE.txt:md5,d4b61d77cb0b810024447774a7d4d141"
                    ]
                ],
                "ectyper_results": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c7fe273c583a801fec7101806fc75d7d"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T18:51:16.412271278"
    }
}